client/x/evmstaking/keeper/msg_queue.go

```diff
@@ -0,0 +1,79 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	slashingtypes "github.com/cosmos/cosmos-sdk/x/slashing/types"
+	stype "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	"github.com/piplabs/story/client/x/evmstaking/types"
+	"github.com/piplabs/story/lib/errors"
+	"github.com/piplabs/story/lib/log"
+)
+
+// EnqueueMsg enqueues a message to the queue of the current epoch.
+func (k Keeper) EnqueueMsg(ctx context.Context, msg types.QueuedMessage) error {
+	return k.MessageQueue.Enqueue(ctx, msg)
+}
+
+// ProcessAllMsgs returns the set of messages queued in a given epoch.
+func (k Keeper) ProcessAllMsgs(ctx context.Context) error {
+	iterator, err := k.MessageQueue.Iterate(ctx)
+	if err != nil {
+		return errors.Wrap(err, "message queue iterator")
+	}
+
+	for ; iterator.Valid(); iterator.Next() {
+		qMsg, err := iterator.Value()
+		if err != nil {
+			return errors.Wrap(err, "get value of message queue")
+		}
+
+		if err := k.processMsg(ctx, &qMsg); err != nil {
+			log.Warn(ctx, "Failed to process queued message", err, "tx_id", string(qMsg.TxId))
+			return errors.Wrap(err, "process queued message")
```

Comment:
I agree that we continue processing messages. I think it would be good to remove warn log here and return error, and just add warn log in `abci.go` and continue processing messages. Since the error returned here is used for our test codes.

Also totally agree that we need to handle failed messages. This is also the part that I had the most trouble with. Since the case where a message fails can be quite complex to consider right now, I'm thinking of storing failed messages separately rather than handling them right away, so that we can handle them later appropriately for example via upgrade. What do you think?